src/core/geometry/qgsgeometry.cpp

```diff
@@ -1746,14 +1746,26 @@ QgsGeometry QgsGeometry::offsetCurve( double distance, int segments, JoinStyle j
   {
     QgsGeos geos( d->geometry.get() );
     mLastError.clear();
-    QgsAbstractGeometry *offsetGeom = geos.offsetCurve( distance, segments, joinStyle, miterLimit, &mLastError );
+
+    // GEOS can flip the curve orientation in some circumstances. So record previous orientation and correct if required
+    const QgsCurve::Orientation prevOrientation = qgsgeometry_cast< const QgsCurve * >( d->geometry.get() )->orientation();
```

Comment:
Maybe I get it wrong here, but isn't orientation() giving meaningless results for non-closed curves?

And why do we need to force orientation here? I am a bit worried that if GEOS does that, then 1. there is a bug in GEOS; or 2. there is some reason why it happens and preserving orientation is not well defined.

Reply:
> Maybe I get it wrong here, but isn't orientation() giving meaningless results for non-closed curves?

In my tests the orientation returned for non-closed curves is consistent, it just doesn't always match with the orientation you'd visually expect. (E.g. a clockwise "horseshoe" curve consistently returns anticlockwise orientation). 

> And why do we need to force orientation here? I am a bit worried that if GEOS does that, then 1. there is a bug in GEOS; or 2. there is some reason why it happens and preserving orientation is not well defined.

This was actually a bit of a dead-end in this PR, and isn't directly related to fixing the rendering issue. But after coding it I thought it would still be desirable in general. E.g. if I use the offset line tool, I'd ideally expect my lines not to be reversed if I offset them by 1 m to the left. 

I believe the underlying issue is by design in geos, at least https://github.com/qgis/QGIS/blob/master/src/core/symbology/qgssymbollayerutils.cpp#L850 , https://issues.qgis.org/issues/13811 lead me to believe this. But I don't think we need/want to follow this same behavior.

And possibly fixing in GEOS is a better approach, but unfortunately we can't rely on GEOS fixes reaching our users until at least 1-2 years (depending on GEOS release cycle and distro policy).

Reply:
If your tests are consistent for non-closed curves then I'm fine - I was mainly reacting with the warning "The result is not predictable for non-closed curves." from API docs in my mind :)

Just thinking out loud, if the trick with reversing linestring on negative offset consistently works, isn't it a better method than orientation()?

Anyway... I trust your judgement, I was only making sure this was intended :-) Feel free to leave as it is.

Reply:
> Just thinking out loud, if the trick with reversing linestring on negative offset consistently works, isn't it a better method than orientation()?

There's two things at play here -- one is the reversed strings which always occur with negative offsets, the other is that even with positive offsets curves can be reversed (depending on their initial orientation). Without this change one of the tests with +ve offset at https://github.com/qgis/QGIS/pull/8424/files#diff-fe3aa1328ee04f0eb00a1b1d59c0ea71R4488 fails.